File: services/analysis.py
import math
import json
from database import db_get, db_all, db_run

def run_analysis(user_id):
    """Multi-Factor Creditworthiness Index (MF-CWI) Algorithm.
    Evaluates credit risk, interest rate, and max approved amount based on 5 dimensions.
    """
    # 1. Fetch data
    profile = db_get('SELECT * FROM client_profiles WHERE user_id = ?', (user_id,))
    if not profile:
        raise ValueError('Client profile not found. Complete profile details first.')

    bank_accounts = db_all('SELECT * FROM bank_accounts WHERE user_id = ?', (user_id,))
    previous_loans = db_all('SELECT * FROM previous_loans WHERE user_id = ?', (user_id,))

    credit_score = profile.get('credit_score', 300)
    annual_income = profile.get('annual_income', 0.0)
    monthly_expenses = profile.get('monthly_expenses', 0.0)
    requested_loan_amount = profile.get('requested_loan_amount', 0.0)

    # 2. Calculations
    monthly_income = annual_income / 12.0 if annual_income > 0 else 0.0

    total_active_monthly_loan_payments = 0.0
    total_remaining_balance = 0.0
    total_original_loan_amount = 0.0

    for loan in previous_loans:
        if loan.get('status') == 'active':
            total_active_monthly_loan_payments += loan.get('monthly_payment', 0.0)
            total_remaining_balance += loan.get('remaining_balance', 0.0)
            total_original_loan_amount += loan.get('loan_amount', 0.0)

    total_reserves = sum(acc.get('balance', 0.0) for acc in bank_accounts)

    # Calculate Debt-to-Income (DTI)
    dti = (monthly_expenses + total_active_monthly_loan_payments) / monthly_income if monthly_income > 0 else 1.0

    # --- Step A: Normalization of Inputs (0 to 1) ---
    # 1. Credit Score (300 to 850)
    s_credit = max(0.0, min(1.0, (credit_score - 300.0) / (850.0 - 300.0)))

    # 2. DTI Score (0 to 60% mapped to 1 to 0)
    s_dti = max(0.0, min(1.0, 1.0 - (dti / 0.60)))

    # 3. Expense-to-Income Score (lower is better, mapped to 0 to 1)
    expense_ratio = monthly_expenses / monthly_income if monthly_income > 0 else 1.0
    s_exp = max(0.0, min(1.0, 1.0 - expense_ratio))

    # 4. Reserves Score (liquid cash vs 3 months of expenses)
    reserve_threshold = max(1000.0, monthly_expenses * 3.0)
    s_res = max(0.0, min(1.0, total_reserves / reserve_threshold))

    # 5. Loan History Score
    s_history = 1.0
    if total_original_loan_amount > 0:
        s_history = max(0.0, min(1.0, 1.0 - (total_remaining_balance / total_original_loan_amount)))

    # --- Step B: Weighted CWI Calculation ---
    cwi = (0.40 * s_credit) + (0.25 * s_dti) + (0.15 * s_exp) + (0.10 * s_res) + (0.10 * s_history)

    # --- Step C: Risk Tiering ---
    risk_score = 'High'
    if cwi >= 0.75:
        risk_score = 'Low'
    elif cwi >= 0.50:
        risk_score = 'Medium'

    # --- Step D: Risk-Adjusted Interest Rate (ROI) ---
    base_rate = 0.045
    risk_premium = 0.15
    roi = base_rate + (1.0 - cwi) * risk_premium

    # --- Step E: Approved Loan Amount Calculation ---
    dti_threshold = 0.30
    if credit_score >= 750:
        dti_threshold = 0.45
    elif credit_score >= 650:
        dti_threshold = 0.38

    max_monthly_payment_allowed = (monthly_income * (dti_threshold * cwi)) - total_active_monthly_loan_payments

    approved_amount = 0.0
    if max_monthly_payment_allowed > 0:
        n = 60
        r = roi / 12.0  # monthly rate
        # Amortization: P = PMT * (1 - (1 + r)^-n) / r
        approved_amount = max_monthly_payment_allowed * (1.0 - math.pow(1.0 + r, -n)) / r

        # Cap at 3x annual income or $250,000 maximum
        cap = min(annual_income * 3.0, 250000.0)
        if approved_amount > cap:
            approved_amount = cap

        # Round to nearest hundred
        approved_amount = round(approved_amount / 100.0) * 100.0

    # --- Step F: Bank Recommendation Matching ---
    bank_policies = [
        {
            'bank': 'Chase Premier Personal',
            'minCreditScore': 720,
            'maxDti': 0.38,
            'baseInterest': 0.0599,
            'maxAmount': 150000,
            'termMonths': 60,
            'description': 'Excellent rates for borrowers with strong credit scores and lower DTI ratio.'
        },
        {
            'bank': 'Bank of America Standard',
            'minCreditScore': 660,
            'maxDti': 0.43,
            'baseInterest': 0.0749,
            'maxAmount': 80000,
            'termMonths': 60,
            'description': 'Competitive options with moderate credit and stable income.'
        },
        {
            'bank': 'Wells Fargo Personal',
            'minCreditScore': 620,
            'maxDti': 0.48,
            'baseInterest': 0.0999,
            'maxAmount': 50000,
            'termMonths': 48,
            'description': 'Flexible terms suited for clients consolidating debts.'
        },
        {
            'bank': 'Capital One Starter',
            'minCreditScore': 580,
            'maxDti': 0.50,
            'baseInterest': 0.1349,
            'maxAmount': 25000,
            'termMonths': 36,
            'description': 'Ideal for establishing credit or shorter term small personal loans.'
        }
    ]

    recommendations = []
    for policy in bank_policies:
        if credit_score >= policy['minCreditScore'] and dti <= policy['maxDti']:
            # Adjust interest rate based on client's CWI
            adjusted_rate = policy['baseInterest'] + (1.0 - cwi) * 0.05
            final_rate = min(adjusted_rate, policy['baseInterest'] + 0.06)

            # Estimate monthly payment
            loan_target = min(policy['maxAmount'], requested_loan_amount, approved_amount or requested_loan_amount)
            mr = final_rate / 12.0
            mt = policy['termMonths']
            est_payment = 0.0
            if loan_target > 0 and mr > 0:
                est_payment = (loan_target * mr) / (1.0 - math.pow(1.0 + mr, -mt))

            recommendations.append({
                'bankName': policy['bank'],
                'offeredRate': round(final_rate * 100.0, 2),
                'maxAmount': policy['maxAmount'],
                'termMonths': policy['termMonths'],
                'estimatedMonthlyPayment': round(est_payment, 2),
                'description': policy['description']
            })

    recommendations.sort(key=lambda x: x['offeredRate'])

    # 3. Save / Update analysis in database
    existing_analysis = db_get('SELECT id FROM loan_analysis WHERE user_id = ?', (user_id,))
    rec_string = json.dumps(recommendations)
    rounded_dti = round(dti, 4)
    rounded_roi = round(roi * 100.0, 2)
    rounded_approved = float(approved_amount)
    rounded_cwi = round(cwi, 4)

    if existing_analysis:
        db_run(
            """UPDATE loan_analysis 
               SET cwi = ?, risk_score = ?, approved_amount = ?, debt_to_income_ratio = ?, interest_rate_offered = ?, recommendations = ?, created_at = CURRENT_TIMESTAMP
               WHERE user_id = ?""",
            (rounded_cwi, risk_score, rounded_approved, rounded_dti, rounded_roi, rec_string, user_id)
        )
    else:
        db_run(
            """INSERT INTO loan_analysis (user_id, cwi, risk_score, approved_amount, debt_to_income_ratio, interest_rate_offered, recommendations)
               VALUES (?, ?, ?, ?, ?, ?, ?)""",
            (user_id, rounded_cwi, risk_score, rounded_approved, rounded_dti, rounded_roi, rec_string)
        )

    return {
        'userId': user_id,
        'cwi': rounded_cwi,
        'riskScore': risk_score,
        'approvedAmount': rounded_approved,
        'debtToIncomeRatio': rounded_dti,
        'interestRateOffered': rounded_roi,
        'recommendations': recommendations
    }

File: services/analysis.js
const { dbGet, dbAll, dbRun } = require('../database');

/**
 * Multi-Factor Creditworthiness Index (MF-CWI) Algorithm
 * A single, highly accurate, mathematical decision algorithm for credit risk,
 * interest rate estimation, and maximum approved loan amount based on 5 dimensions.
 */
const runAnalysis = async (userId) => {
  // 1. Fetch data
  const profile = await dbGet('SELECT * FROM client_profiles WHERE user_id = ?', [userId]);
  if (!profile) {
    throw new Error('Client profile not found. Complete profile details first.');
  }

  const bankAccounts = await dbAll('SELECT * FROM bank_accounts WHERE user_id = ?', [userId]);
  const previousLoans = await dbAll('SELECT * FROM previous_loans WHERE user_id = ?', [userId]);

  const { credit_score, annual_income, monthly_expenses, requested_loan_amount } = profile;

  // 2. Calculations
  const monthlyIncome = annual_income / 12;

  // Sum of active previous loan monthly payments and remaining balances
  let totalActiveMonthlyLoanPayments = 0;
  let totalRemainingBalance = 0;
  let totalOriginalLoanAmount = 0;

  previousLoans.forEach((loan) => {
    if (loan.status === 'active') {
      totalActiveMonthlyLoanPayments += loan.monthly_payment;
      totalRemainingBalance += loan.remaining_balance;
      totalOriginalLoanAmount += loan.loan_amount;
    }
  });

  // Sum of bank account balances
  const totalReserves = bankAccounts.reduce((sum, acc) => sum + acc.balance, 0);

  // Calculate Debt-to-Income (DTI)
  // DTI = (Monthly Expenses + Active Loan Payments) / Monthly Income
  const dti = monthlyIncome > 0 
    ? (monthly_expenses + totalActiveMonthlyLoanPayments) / monthlyIncome 
    : 1.0;

  // --- Step A: Normalization of Inputs (0 to 1) ---
  // 1. Credit Score (300 to 850)
  const sCredit = Math.max(0, Math.min(1, (credit_score - 300) / (850 - 300)));

  // 2. DTI Score (0 to 60% mapped to 1 to 0)
  const sDti = Math.max(0, Math.min(1, 1 - (dti / 0.60)));

  // 3. Expense-to-Income Score (lower is better, mapped to 0 to 1)
  const expenseRatio = monthlyIncome > 0 ? (monthly_expenses / monthlyIncome) : 1.0;
  const sExp = Math.max(0, Math.min(1, 1 - expenseRatio));

  // 4. Reserves Score (liquid cash vs 3 months of expenses)
  const reserveThreshold = Math.max(1000, monthly_expenses * 3);
  const sRes = Math.max(0, Math.min(1, totalReserves / reserveThreshold));

  // 5. Loan History Score
  let sHistory = 1.0;
  if (totalOriginalLoanAmount > 0) {
    // Score increases as remaining balance decreases relative to original loan
    sHistory = Math.max(0, Math.min(1, 1 - (totalRemainingBalance / totalOriginalLoanAmount)));
  }

  // --- Step B: Weighted CWI Calculation ---
  // Weights: Credit Score (40%), DTI (25%), Expense Ratio (15%), Cash Reserves (10%), Loan History (10%)
  const cwi = (0.40 * sCredit) + (0.25 * sDti) + (0.15 * sExp) + (0.10 * sRes) + (0.10 * sHistory);

  // --- Step C: Risk Tiering ---
  let riskScore = 'High';
  if (cwi >= 0.75) {
    riskScore = 'Low';
  } else if (cwi >= 0.50) {
    riskScore = 'Medium';
  }

  // --- Step D: Risk-Adjusted Interest Rate (ROI) ---
  // Base Rate = 4.5%, Risk Premium = 15%. ROI = Base + (1 - CWI) * Premium
  const baseRate = 0.045;
  const riskPremium = 0.15;
  const roi = baseRate + (1 - cwi) * riskPremium;

  // --- Step E: Approved Loan Amount Calculation ---
  // Max allowable monthly payment based on Credit Score tier and CWI
  let dtiThreshold = 0.30;
  if (credit_score >= 750) {
    dtiThreshold = 0.45;
  } else if (credit_score >= 650) {
    dtiThreshold = 0.38;
  }

  const maxMonthlyPaymentAllowed = (monthlyIncome * (dtiThreshold * cwi)) - totalActiveMonthlyLoanPayments;

  let approvedAmount = 0;
  if (maxMonthlyPaymentAllowed > 0) {
    // Amortization formula for 60-month term (5 years)
    const n = 60;
    const r = roi / 12; // monthly rate
    // P = PMT * (1 - (1 + r)^-n) / r
    approvedAmount = maxMonthlyPaymentAllowed * (1 - Math.pow(1 + r, -n)) / r;

    // Cap at 3x annual income or $250,000 maximum
    const cap = Math.min(annual_income * 3, 250000);
    if (approvedAmount > cap) {
      approvedAmount = cap;
    }
    // Round to nearest hundred
    approvedAmount = Math.round(approvedAmount / 100) * 100;
  }

  // --- Step F: Bank Recommendation Matching ---
  const bankPolicies = [
    {
      bank: 'Chase Premier Personal',
      minCreditScore: 720,
      maxDti: 0.38,
      baseInterest: 0.0599,
      maxAmount: 150000,
      termMonths: 60,
      description: 'Excellent rates for borrowers with strong credit scores and lower DTI ratio.'
    },
    {
      bank: 'Bank of America Standard',
      minCreditScore: 660,
      maxDti: 0.43,
      baseInterest: 0.0749,
      maxAmount: 80000,
      termMonths: 60,
      description: 'Competitive options with moderate credit and stable income.'
    },
    {
      bank: 'Wells Fargo Personal',
      minCreditScore: 620,
      maxDti: 0.48,
      baseInterest: 0.0999,
      maxAmount: 50000,
      termMonths: 48,
      description: 'Flexible terms suited for clients consolidating debts.'
    },
    {
      bank: 'Capital One Starter',
      minCreditScore: 580,
      maxDti: 0.50,
      baseInterest: 0.1349,
      maxAmount: 25000,
      termMonths: 36,
      description: 'Ideal for establishing credit or shorter term small personal loans.'
    }
  ];

  // Recommendations mapping
  const recommendations = bankPolicies
    .filter((policy) => credit_score >= policy.minCreditScore && dti <= policy.maxDti)
    .map((policy) => {
      // Adjust interest rate slightly based on client's CWI
      // Higher CWI = closer to the base rate of the bank policy
      const adjustedRate = policy.baseInterest + (1 - cwi) * 0.05;
      const finalRate = Math.min(adjustedRate, policy.baseInterest + 0.06);

      // Estimate monthly payment for the bank's maximum possible loan or user's requested amount
      const loanTarget = Math.min(policy.maxAmount, requested_loan_amount, approvedAmount || requested_loan_amount);
      const mr = finalRate / 12;
      const mt = policy.termMonths;
      const estPayment = loanTarget > 0 
        ? (loanTarget * mr) / (1 - Math.pow(1 + mr, -mt))
        : 0;

      return {
        bankName: policy.bank,
        offeredRate: parseFloat((finalRate * 100).toFixed(2)),
        maxAmount: policy.maxAmount,
        termMonths: policy.termMonths,
        estimatedMonthlyPayment: parseFloat(estPayment.toFixed(2)),
        description: policy.description
      };
    })
    .sort((a, b) => a.offeredRate - b.offeredRate);

  // 3. Save / Update analysis in database
  const existingAnalysis = await dbGet('SELECT id FROM loan_analysis WHERE user_id = ?', [userId]);

  const recString = JSON.stringify(recommendations);
  const roundedDti = parseFloat(dti.toFixed(4));
  const roundedRoi = parseFloat((roi * 100).toFixed(2));
  const roundedApproved = parseFloat(approvedAmount.toFixed(2));
  const roundedCwi = parseFloat(cwi.toFixed(4));

  if (existingAnalysis) {
    await dbRun(
      `UPDATE loan_analysis 
       SET cwi = ?, risk_score = ?, approved_amount = ?, debt_to_income_ratio = ?, interest_rate_offered = ?, recommendations = ?, created_at = CURRENT_TIMESTAMP
       WHERE user_id = ?`,
      [roundedCwi, riskScore, roundedApproved, roundedDti, roundedRoi, recString, userId]
    );
  } else {
    await dbRun(
      `INSERT INTO loan_analysis (user_id, cwi, risk_score, approved_amount, debt_to_income_ratio, interest_rate_offered, recommendations)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [userId, roundedCwi, riskScore, roundedApproved, roundedDti, roundedRoi, recString]
    );
  }

  return {
    userId,
    cwi: parseFloat(cwi.toFixed(4)),
    riskScore,
    approvedAmount: roundedApproved,
    debtToIncomeRatio: roundedDti,
    interestRateOffered: roundedRoi,
    recommendations
  };
};

module.exports = {
  runAnalysis
};
